File: schema/toolpack.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, Field

from schema.action import RiskLevel


class ExecutionErrorCode(StrEnum):
    CONNECT_FAILED = "connect_failed"
    AUTH_FAILED = "auth_failed"
    HOST_KEY_CHANGED = "host_key_changed"
    PERMISSION_DENIED = "permission_denied"
    POLICY_DENIED = "policy_denied"
    APPROVAL_REQUIRED = "approval_required"
    TIMEOUT = "timeout"
    CANCELED = "canceled"
    PROCESS_FAILED = "process_failed"
    OUTPUT_TRUNCATED = "output_truncated"
    TOOL_MISSING = "tool_missing"
    PLATFORM_UNSUPPORTED = "platform_unsupported"
    SCOPE_DENIED = "scope_denied"


class ExecutionArtifact(BaseModel):
    id: str
    path: str
    media_type: str = "text/plain"
    size: int = Field(ge=0)


class ExecutionResult(BaseModel):
    ok: bool
    execution_id: str
    summary: str
    structured: dict[str, Any] = Field(default_factory=dict)
    artifact_refs: list[ExecutionArtifact] = Field(default_factory=list)
    exit_code: int | None = None
    started_at: datetime
    finished_at: datetime
    truncated: bool = False
    error_code: ExecutionErrorCode | None = None


class ToolBackend(StrEnum):
    LOCAL = "local"
    SSH = "ssh"


class ToolRunStatus(StrEnum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELED = "canceled"


class ToolManifestSchema(BaseModel):
    id: str = Field(min_length=1, max_length=128)
    name: str = Field(min_length=1, max_length=128)
    description: str = ""
    backend: ToolBackend
    executable: str = Field(min_length=1, max_length=128)
    category: str = Field(min_length=1, max_length=64)
    action_type: str = Field(min_length=3, max_length=128)
    risk_level: RiskLevel = RiskLevel.L1
    default_timeout_seconds: int = Field(gt=0)
    max_timeout_seconds: int = Field(gt=0)
    input_schema: dict[str, Any] = Field(default_factory=dict)
    output_schema: dict[str, Any] = Field(default_factory=dict)
    policy: dict[str, Any] = Field(default_factory=dict)


class ToolSchema(BaseModel):
    id: str
    name: str
    description: str
    backend: ToolBackend
    category: str
    available: bool | None
    availability_message: str = ""
    install_hint: str = ""
    manifest: ToolManifestSchema


class QueryToolpackToolsResponse(BaseModel):
    tools: list[ToolSchema]


class ToolRunRequest(BaseModel):
    sandbox_container_id: int = Field(gt=0)
    input: dict[str, Any] = Field(default_factory=dict)
    timeout_seconds: int | None = Field(default=None, gt=0)


class ToolRunSnapshot(BaseModel):
    run_id: str
    tool_id: str
    sandbox_container_id: int
    status: ToolRunStatus
    result: ExecutionResult | None = None
    started_at: datetime
    finished_at: datetime | None = None


class ToolRunCancelResponse(BaseModel):
    run_id: str
    canceled: bool
    status: Literal["running", "completed", "failed", "canceled"]

File: web/src/shared/api/toolpack.ts
import { apiGet, apiPost } from "./client";
import { buildQuery } from "./query";
import type { CommonResponsePayload } from "./types";

export type ToolBackend = "local" | "ssh";
export type ToolRunStatus = "running" | "completed" | "failed" | "canceled";
export type JsonSchemaProperty = {
  type?: string | string[];
  title?: string;
  description?: string;
  default?: unknown;
  minimum?: number;
  maximum?: number;
  minLength?: number;
  maxLength?: number;
  enum?: unknown[];
};

export type ToolInputSchema = {
  type?: string;
  required?: string[];
  properties?: Record<string, JsonSchemaProperty>;
  additionalProperties?: boolean;
};

export type ToolManifest = {
  id: string;
  name: string;
  description: string;
  backend: ToolBackend;
  executable: string;
  category: string;
  action_type: string;
  risk_level: string;
  default_timeout_seconds: number;
  max_timeout_seconds: number;
  input_schema: ToolInputSchema;
  output_schema: Record<string, unknown>;
  policy: Record<string, unknown>;
};

export type ToolSchema = {
  id: string;
  name: string;
  description: string;
  backend: ToolBackend;
  category: string;
  available: boolean | null;
  availability_message: string;
  install_hint: string;
  manifest: ToolManifest;
};

export type ExecutionArtifact = {
  id: string;
  path: string;
  media_type: string;
  size: number;
};

export type ExecutionResult = {
  ok: boolean;
  execution_id: string;
  summary: string;
  structured: {
    stdout?: string;
    records?: unknown[];
    [key: string]: unknown;
  };
  artifact_refs: ExecutionArtifact[];
  exit_code: number | null;
  started_at: string;
  finished_at: string;
  truncated: boolean;
  error_code: string | null;
};

export type ToolRunSnapshot = {
  run_id: string;
  tool_id: string;
  sandbox_container_id: number;
  status: ToolRunStatus;
  result: ExecutionResult | null;
  started_at: string;
  finished_at: string | null;
};

export type QueryToolpackToolsResponse = CommonResponsePayload & {
  data?: { tools: ToolSchema[] } | null;
};

export type ToolRunResponse = CommonResponsePayload & {
  data?: ToolRunSnapshot | null;
};

export type ToolRunCancelResponse = CommonResponsePayload & {
  data?: { run_id: string; canceled: boolean; status: ToolRunStatus } | null;
};

const TOOLPACK_PATH = "/api/toolpack";

export function listToolpackTools(sandboxContainerId?: number) {
  return apiGet<QueryToolpackToolsResponse>(`${TOOLPACK_PATH}/tools${buildQuery({ sandbox_container_id: sandboxContainerId })}`);
}

export function startToolRun(toolId: string, payload: {
  sandbox_container_id: number;
  input: Record<string, unknown>;
  timeout_seconds?: number;
}) {
  return apiPost<ToolRunResponse>(`${TOOLPACK_PATH}/tools/${encodeURIComponent(toolId)}/runs`, payload);
}

export function getToolRun(runId: string) {
  return apiGet<ToolRunResponse>(`${TOOLPACK_PATH}/runs/${encodeURIComponent(runId)}`);
}

export function cancelToolRun(runId: string) {
  return apiPost<ToolRunCancelResponse>(`${TOOLPACK_PATH}/runs/${encodeURIComponent(runId)}/cancel`);
}

export function buildToolArtifactUrl(artifactId: string) {
  return `${TOOLPACK_PATH}/artifacts/${encodeURIComponent(artifactId)}`;
}
